File: Models/Occupancy.swift
import SwiftData

@Model
final class Occupancy {
    var id: UUID
    var cropName: String
    var variety: String?
    var startDate: Date
    var endDate: Date?
    var expectedHarvestDate: Date?
    var expectedReleaseDate: Date?
    var status: OccupancyStatus
    var notes: String?
    var createdAt: Date
    var updatedAt: Date

    @Relationship(deleteRule: .cascade, inverse: \GrowingSpace.occupancies)
    var growingSpace: GrowingSpace?

    @Relationship(deleteRule: .nullify, inverse: \Plant.occupancy)
    var plant: Plant?

    @Relationship(deleteRule: .nullify, inverse: \PlannedPlanting.occupancy)
    var plannedPlanting: PlannedPlanting?

    init(
        cropName: String,
        variety: String? = nil,
        startDate: Date,
        endDate: Date? = nil,
        expectedHarvestDate: Date? = nil,
        expectedReleaseDate: Date? = nil,
        status: OccupancyStatus = .active,
        notes: String? = nil,
        growingSpace: GrowingSpace? = nil,
        plant: Plant? = nil
    ) {
        self.id = UUID()
        self.cropName = cropName
        self.variety = variety
        self.startDate = startDate
        self.endDate = endDate
        self.expectedHarvestDate = expectedHarvestDate
        self.expectedReleaseDate = expectedReleaseDate
        self.status = status
        self.notes = notes
        self.growingSpace = growingSpace
        self.plant = plant
        self.createdAt = Date()
        self.updatedAt = Date()
    }

    var displayName: String {
        if let variety = variety, !variety.isEmpty {
            return "\(variety) \(cropName)"
        }
        return cropName
    }
}

enum OccupancyStatus: String, CaseIterable {
    case active
    case completed
    case cancelled

    var displayName: String {
        switch self {
        case .active: return "Active"
        case .completed: return "Completed"
        case .cancelled: return "Cancelled"
        }
    }
}

File: app/src/main/java/com/soilandsupper/domain/model/Occupancy.kt
package com.soilandsupper.domain.model

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.PrimaryKey

@Entity(
    tableName = "occupancies",
    foreignKeys = [
        ForeignKey(
            entity = GrowingSpace::class,
            parentColumns = ["id"],
            childColumns = ["growingSpaceId"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = Plant::class,
            parentColumns = ["id"],
            childColumns = ["plantId"],
            onDelete = ForeignKey.SET_NULL
        ),
        ForeignKey(
            entity = PlannedPlanting::class,
            parentColumns = ["id"],
            childColumns = ["plannedPlantingId"],
            onDelete = ForeignKey.SET_NULL
        )
    ]
)
data class Occupancy(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val cropName: String,
    val variety: String? = null,
    val startDate: Long,
    val endDate: Long? = null,
    val expectedHarvestDate: Long? = null,
    val expectedReleaseDate: Long? = null,
    val status: String = OccupancyStatus.ACTIVE.name,
    val notes: String? = null,
    val growingSpaceId: Long? = null,
    val plantId: Long? = null,
    val plannedPlantingId: Long? = null,
    val createdAt: Long = System.currentTimeMillis(),
    val updatedAt: Long = System.currentTimeMillis()
) {
    val displayName: String
        get() = if (!variety.isNullOrBlank()) "$variety $cropName" else cropName
}

enum class OccupancyStatus {
    ACTIVE,
    COMPLETED,
    CANCELLED;

    val displayName: String
        get() = when (this) {
            ACTIVE -> "Active"
            COMPLETED -> "Completed"
            CANCELLED -> "Cancelled"
        }
}
